src/org/rascalmpl/library/String.rsc

```diff
@@ -680,4 +682,124 @@ or the indentation.
 * This function works fine if `indentation` is not spaces or tabs; but it does not make much sense. 
 }
 @javaClass{org.rascalmpl.library.Prelude}
-java str indent(str indentation, str content, bool indentFirstLine=false);
\ No newline at end of file
+java str indent(str indentation, str content, bool indentFirstLine=false);
+
+list[str] newLineCharacters = [
+    "\u000A", // LF
+    "\u000B", // VT
+    "\u000C", // FF
+    "\u000D", // CR
+    "\u000D\u000A", // CRLF
+    "\u0085", // NEL
+    "\u2028", // LS
+    "\u2029" // PS
+];
+
+@synopsis{Comparator to sort strings by length (ascending).}
+private bool bySize(str a, str b) = size(a) < size(b);
+
+@synopsis{Comparator to sort strings by relative position in a reference list.}
+private bool(str, str) byIndex(list[str] indices) {
+    return bool(str a, str b) {
+        return indexOf(indices, a) < indexOf(indices, b);
+    };
+}
+
+@synopsis{Determine the most-used newline character in a string.}
+str mostUsedNewline(str input, list[str] lineseps = newLineCharacters, str(list[str]) tieBreaker = getFirstFrom) {
+    linesepCounts = (nl: 0 | nl <- lineseps);
+    for (nl <- sort(lineseps, bySize)) {
+        int count = size(findAll(input, nl));
+        linesepCounts[nl] = count;
+        // subtract all occurrences of substrings of newline characters that we counted before
+        for (str snl <- substrings(nl), linesepCounts[snl]?) {
```

Comment:
this almost looks like pattern matching on strings? (which we only have reasonable support over)

for example:

```
rascal>visit("abcd") { case str m : println(m); }
abcd
bcd
cd
d
```


come to think of it, this whole function smells like an parsing automata. Where we build a big state table of all the possible matches and then iterate through all the chars and count the matches based on their state.

In java this would be 20/30 lines, but in rascal we might be missing some primitives (as we don't have a character loop).

Reply:
If we hard-code the set of newline characters (e.g. to all Unicode newline chars), we could write it as a grammar and use the parser generator. Downside is that all (transitive) imports of this module will trigger generation of a parser. We could also move some of those to a specific `Format` module.